File: fcfs.py
# -*- coding: utf-8 -*-
"""
Created on Mon Sep  2 03:07:39 2019

@author: Nitin
"""
#first come first serve scheduler
'''
To a CPU, the processes shall come in the order of their arrival time by default. 
Sorting is not necessary
'''

at = []
wt = []
tat = []
bt = []

num = int(input("Enter the number of processes: "))
for i in range(num):
    print("\nEnter the arrival time of process",i+1,":")
    p = eval(input())
    at.append(p)


for i in range(num):
    print("\nEnter burst time of process",i+1,":" )
    m = eval(input())
    bt.append(m)
    
#however, if sorting is to be done,
'''for i in range(num):
    pos = i
    for j in range(i+1,num):
        if (at[i]<at[j]):
            pos=j
        else:
            temp = at[i]
            at[i]=at[pos]
            at[pos] = temp
            
            temp = bt[i]
            bt[i]=bt[pos]
            bt[pos] = temp
            
            temp = pt[i]     //process name
            pt[i]=pt[pos]
            pt[pos] = temp
'''
     
        
wt.append(0)
for i in range(1,num):
    w = bt[i-1]+wt[i-1]
    wt.append(w)
for i in range(num):
    t = bt[i]+wt[i]
    tat.append(t)
awt=sum(wt)/num
atat = sum(at)/num  
print("process\t arrival-time\t wait-time\tturnaroundtime\n")
for i in range(num):
    print("process",i+1,"\t",at[i],"\t\t ",wt[i],"\t\t", tat[i])
print("average wait time is: ",awt)
print("\naverage turnaroundtime is: ",atat)
print("\n")

File: first come first serve/fcfs.cpp
//This program is to generate fcfs algorithm
//It is based on the premise that the order in which CPU receives programs is the order in which it records arrival time
//So, sorting is not actually needed in a fcfs algorithm
#include<stdio.h>
int main()
{
	int wait=0,turn=0,at[10], wt[10], tat[10], bt[10], temp=0, pos=0,i,j, num; 
	float awt = 0, atat = 0; 
	//or one can come up with dynamic allocation
	/*the arrival time of the processes are recorded by the cpu in the order they arrive*/
    printf("Enter the number of processes to be entered: ");
    scanf("%d",&num);
    for (i=0;i<num;i++)
    {
    	printf("\nEnter the arrival time of the processes %d: ",i+1);
    	scanf("%d",&at[i]);
    	printf("\nEnter the burst time of the process: ");
    	scanf("%d",&bt[i]);
	}
	wt[0] = 0;  //waiting time for the first process is zero.
	for(j=1;j<num;j++)
	{
		wt[j] = bt[j-1]+wt[j-1];
		wait +=wt[j]; 
	}
	for(i=0;i<num;i++)
	{
		tat[i] = wt[i] + bt[i];
		turn +=tat[i];
	}
	atat = (float)turn/num;
	awt = (float)wait/num;
	printf("The final table is hence: \n process\t arrival_time\t wait_time\t turn_around_time\n");
	for(i=0;i<num;i++)
	{
		printf("process %d\t\t %d\t\t %d\t\t %d\n",i+1,at[i],wt[i],tat[i]);
	}
	printf("The average turnaroundtime is: %f and the average wait time is: %f\n",atat,awt);
	return 0;
}
